### ISSUE TITLE
Use map data in systems

### ISSUE THREAD

**Issue by MTRNord** [katharostech/bevy_ldtk#6]
Hi :) How do I access the map from a system? I am new to bevy so I might have missed something. My first try was to use ` map: Query<LdtkMap>` but that didnt work. 

Basicly I want to get from the map the entities to draw them as needed. And the required information is stored in the map.

**Comment by dannytatom** [comment]
I was looking for the same thing, but I don't think it's supported. It looks like any layer without a tileset attached [is skipped over](https://github.com/katharostech/bevy_ldtk/blob/9e76f0d66ace16600f3f307901bb5e7abd5dce89/src/system.rs#L135-L138).

**Comment by zicklag** [comment]
Hi there!

You were very close with `Query<&LdtkMap>`, you actually have to use `Query<&Handle<LdtkMap>>`. I'll have to add this to the docs apparently. 😃 Thanks for asking. :+1:

```rust
pub fn setup_collisions(
    map_query: Query<&Handle<LdtkMap>>,
    map_assets: Res<Assets<LdtkMap>>,
) {
    // Loop through all of the maps
    for map_handle in map_query.iter() {
        // We have to `if let` here because asset server may not have finished loading
        // the map yet.
        if let Some(map) = map_assets.get(map_handle) {
            // Do something with map data
        }
    }
}
```

At that point you have full access to the whole map data structure which you can do whatever you want with.

**Comment by MTRNord** [comment]
awesome! yeah having that in the example would have helped :)

**Comment by zicklag** [comment]
You can leave the issue open and I'll close it when I get a good example in. :smile:  :+1:

**Comment by MTRNord** [comment]
@zicklag if you have the time and want to: An example to get the coordinates of the entities using this would also be nice. I did get the texture loaded as wanted but I am kinda stuck wrapping my head around how to find the x,y coordinates for the entity :) I am probably just missing something again. (aka do i need Layer defs or layer instances, or both and how to bring them into the bevy coordinate format)

**Comment by zicklag** [comment]
There may be minor typos in the below example, but heres a modification of the actual code I've used in my game for spawning the player:

```rust
if let Some(map) = maps.get(handle) {
    // This is the default level, but if you spawned a different level, put that ID here
    let level_idx = 0;

    // Get the level from the project
    let level = &map.project.levels[level_idx];

    /// Find the entities layer
    let entities_layer = level
        .layer_instances
        .as_ref() // get a reference to the layer instances
        .unwrap() // Unwrap the option ( this could be None, if there are no layers )
        .iter() // Iterate over the layers
        .filter(|&x| x.__identifier == "MyEntities") // Filter on the one we want
        .next() // Get it
        .unwrap() // Unwrap it ( would be None if it could not find a layer "MyEntities" );


    // Get the specific entity you want
    let player_start = entities_layer
        .entity_instances
        .iter() // Iterate over our entities in the layer
        .filter(|x| x.__identifier == "PlayerStart") // Find the one we want
        .next() // Get it
        .unwrap() // Unwrap it;

    // Get the number of layers in the map and add one to it: this is how high we need to
    // spawn the player so that he is on top of all the maps
    let player_z = level.layer_instances.as_ref().unwrap().len() as f32 + 1.0;

    // Spawn the entity!
    commands
        .spawn(SpriteBundle {
            // Set your sprite stuff
            transform: Transform::from_xyz(
                // The player x position is the entity's x position from the map data
                player_start.px[0],
                // The player y position is the entity's y position from the map data, but
                // multiplied by negative one because in the LDtk map +y means down and not up.
                player_start.px[1] * -1,
                // Spawn the player with the z value we determined earlier
                player_z
            ),
            ..Default::default()
        });
    }
}
```

It's a little wordier than I would like: the LDtk map format could probably use some helpers to make this kind of stuff a little easier, but at least for now all the data is there. If me or my team get the chance sometime we might try to add some helpers.

**Comment by MTRNord** [comment]
Hm I guess it should work but for some reason it seems like it ends up not where it is in the editor.

Editor:

![grafik](https://user-images.githubusercontent.com/1374914/111230399-dc022600-85e7-11eb-867c-8ee5057b1a1e.png)


Game:

![grafik](https://user-images.githubusercontent.com/1374914/111230419-e3293400-85e7-11eb-85f9-f3768ec644c1.png)


(And yes due to laziness and learning purposes I am using one of the starter levels the editor brought)

**Comment by zicklag** [comment]
Ah, do you have the `center_map` option set to `true` in when you spawn the map? That will cause the coordinates to be off. The setting is really more of a convenience for when you just want to display the map.

I guess that's a little confusing maybe. :thinking: maybe I shouldn't have the option to avoid confusion? But then maybe people would be annoyed that they couldn't spawn the map without it. I'm open to suggestions on that point. :slightly_smiling_face: 

---

> (And yes due to laziness and learning purposes I am using one of the starter levels the editor brought)

That ain't lazyness that's efficiency! 😉

**Comment by MTRNord** [comment]
Ah yeah :) though I think the `coordinate_at_mouse` example thingy by the engine has a conversion for calculating the coordinates based on the window size. That should move it in the right place right?

**Comment by zicklag** [comment]
Hmm I haven't seen that before. I don't know. If you want it centered or something during gameplay you would most-likely just move the camera to follow the player or something like that.

**Comment by MTRNord** [comment]
Intresting enough without the `center_map` set to true it is different wrong.

![grafik](https://user-images.githubusercontent.com/1374914/111231387-757e0780-85e9-11eb-8896-4749f8bf22b8.png)


I guess it might be due to using a 4k monitor :)

**Comment by zicklag** [comment]
Hmm, that's interesting. Is the map scale set to 1? I _think_ the monitor scale shouldn't make a difference.

**Comment by MTRNord** [comment]
Ahhh thats it. It was set to 4. Am I right to assume multiplying each coordinate with the scale factor will fix the issue? (It works with scale 1 but scale 1 doesnt cover the screen the way I want it too)

**Comment by zicklag** [comment]
Yes, you could multiply everything by 4 and I think it would work, but there's a much more awesome way to do it that will work no matter what your screen size. You just have to spawn the camera with different settings that tell it to fix the height of the camera and then automatically reveal the rest ofthe scene along the width of the camera, and to also scale the whole scene up if the height of the screen is bigger than the height of the screen:

```rust
commands
                .spawn(OrthographicCameraBundle {
                    orthographic_projection: OrthographicProjection {
                        far: 1024.0, // This gives us 1024 layers,
                        scale: 200, // How many pixels high in the game
                        scaling_mode: ScalingMode::FixedVertical, 
                        ..Default::default()
                    },
                    ..OrthographicCameraBundle::new_2d()
                })
```

**Comment by MTRNord** [comment]
Oh yeah thats much cooler :D Thanks for all the help!

**Comment by MTRNord** [comment]
Ah but that doesnt work in 0-4 apparently due to missing scale and scaling_mode fields :D I guess I am now movin to the latest version now

**Comment by zicklag** [comment]
Oh, yeah, I forgot about that. 0.5 should be comming out soon-ish I think so looking forward to that. :)

> Oh yeah thats much cooler :D Thanks for all the help!

Glad I could help 😃

**Comment by MTRNord** [comment]
Just had to figure out dep hell  :D 

Side note: This breaks with 0.5 after bevy commit `079b3ade89195ef1aa4bcf1f8c0740a6572b84b6` (thats the las working one or me)

**Comment by zicklag** [comment]
Ah, gotcha. Good job finding the working one! After 0.5 is released I'll make a new `bevy_ldtk` release that works.

**Comment by MTRNord** [comment]
Yeah thats fine :D No rush :3 Learned today more in this issue than in the whole university semester ;P

**Comment by zicklag** [comment]
Fixed in 173ecbd14157423e45327c451000c8dc6bb0eddf.
